### PR TITLE
Send email notifications for changes to "task" measures

### LINKED ISSUE #49: Task updated: Set up email notifications for changes to measures of type "task" (measuretype_id: 5)
We need to send email notifications to users assigned to a task when that task is updated. Ideally users should not get multiple emails for the same task if the changes have happened seemingly at the same time (within a minute, see comment below for discussion of options)

Specifically, emails should be sent when there are changes to 

- any of the measures attributes except `draft`
- any of the measure relationships (create, update, delete), specifically
  - actor_measures 
  - measure_actors 
  - measure_categories (create only, not updateable)
  - measure_indicators (create only, not updateable)
  - measure_measures (create only, not updateable)
  - measure_resources (to be confirmed)
  - user_measures (to be confirmed)

for all measures 

- where `measuretype_id  = 5` ("tasks", should be configurable)
- where `draft = false` (before AND after any update)

to all **assigned users**, i.e. that 

- have a relationship to that task via the `user_measures` table
- are not the same user updating the database as specified above

#### Email: 

Emails should contain link to the measure (will require storing the client-side path in the server configuration)

### PR DESCRIPTION
**Pull request by parndt**
Resolves #49 

We need to send email notifications to users assigned to a task when that task is updated. 

Specifically, emails should be sent when there are changes to:

- [x] any of the measures attributes except `draft`
- [x] any of the measure relationships (create, update, delete), specifically
  - [x] actor_measures 
  - [x] measure_actors 
  - [x] measure_categories (create only, not updateable)
  - [x] measure_indicators (create only, not updateable)
  - [x] measure_measures (create only, not updateable)
  - [x] measure_resources (to be confirmed)
  - [x] user_measures (to be confirmed)

for all measures 

- [x] where `measuretype_id.notifications = true` ("tasks")
- [x] where `draft = false` (before AND after any update)

to all **assigned users**, i.e. that 

- [x] have a relationship to that task via the `user_measures` table
- [x] are not the same user updating the database as specified above

#### Email: 

Emails contain a link to the measure.

### CODE REVIEW

**Comment by parndt** [comment]
@tmfrnz what about for `is_archive` measures? Do we send notifications for them?

**Comment by tmfrnz** [comment]
> @tmfrnz what about for `is_archive` measures? Do we send notifications for them?

I guess we will want notifications when `is_archive` changes but will check with client if we want this when other attributes are changed while it was true

**Comment by parndt** [comment]
>measure_categories (create only, not updateable)
>measure_indicators (create only, not updateable)
>measure_measures (create only, not updateable)

Do you want deletes to also trigger this, or just creates?

**Comment by tmfrnz** [comment]
> > measure_categories (create only, not updateable)
> > measure_indicators (create only, not updateable)
> > measure_measures (create only, not updateable)
> 
> Do you want deletes to also trigger this, or just creates?

yes please! that would be awesome. thanks

**Comment by tmfrnz** [comment]
> > @tmfrnz what about for `is_archive` measures? Do we send notifications for them?
> 
> I guess we will want notifications when `is_archive` changes but will check with client if we want this when other attributes are changed while it was true

@parndt no need to send notifications when a task is archived, unless it is the "is_archive" field that changes (so we want assigned users to know if a task was archived or un-archived but not if there were any updates in the meantime)

**Comment by parndt** [comment]
@tmfrnz this one's a complex one 😁 nearly there, the remaining item is sending the email when a relationship is changed.

**Comment by tmfrnz** [comment]
> @tmfrnz this one's a complex one 😁 nearly there, the remaining item is sending the email when a relationship is changed.

@parndt there is probably more to it but I thought it is simply a matter of responding to changes to the new "relationship_updated_at" field?

**Comment by parndt** [comment]
@tmfrnz I figured out a good compromise; code to follow.

**Comment by parndt** [comment]
@tmfrnz code updated - now it will send an email whenever `relationship_updated_at` (introduced in #46) changes and it meets the other `draft`/`is_archive` criteria.

Please can you have a look through and let me know if it makes sense - you can use the specs for this.

File: app/models/measure.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class Measure < VersionedRecord
+  TASK_MEASURETYPE_ID = 5
```
Comment (tmfrnz) [inline app/models/measure.rb:4]: this should probably be better stored elsewhere

or maybe even better store this in a new field in the "measuretypes" table? e.g. `t.boolean send_email_on_update`

Comment (parndt) [reply app/models/measure.rb:4]: ooh yeah stored in a column would make way more sense!!

File: app/models/measure.rb [hunk 2]
```diff
@@ -49,10 +67,12 @@ def different_parent
...
 
-  def parent_id_allowed_by_measuretype
-    if parent_id && !parent.measuretype&.has_parent
-      errors.add(:parent_id, "is not allowed for this measuretype")
-    end
+  def notify?
+    task? &&
+      notifications? &&
+      (!draft? && !saved_change_to_attribute?(:draft)) &&
+      (!is_archive || saved_change_to_attribute?(:is_archive)) &&
```
Comment (tmfrnz) [inline app/models/measure.rb:74]: should this be `is_archive?` (with the '?')

File: spec/models/measure_spec.rb [hunk 3]
```diff
@@ -95,4 +95,49 @@
...
+      it "won't send when relationship_updated_at changes" do
+        expect { subject.touch(:relationship_updated_at) }
+          .not_to change { ActionMailer::Base.deliveries.count }.from(0)
+      end
+    end
+
+    context "for 'task' measures" do
+      before { allow(subject).to receive(:task?).and_return(true) }
+
+      context "when the current user has the user_measure" do
```
Comment (tmfrnz) [inline spec/models/measure_spec.rb:119]: not sure I understand this test case

File: app/models/measure.rb (hunk 2 above)
Comment (parndt) [reply app/models/measure.rb:74]:
```suggestion
      (!is_archive? || saved_change_to_attribute?(:is_archive)) &&
```

Yes it should be

File: spec/models/measure_spec.rb (hunk 3 above)
Comment (parndt) [inline spec/models/measure_spec.rb:119]:
```suggestion
      context "when the current user owns the task" do
```

Does this make more sense? Basically it's guarding against this case:

> are not the same user updating the database as specified above

Comment (tmfrnz) [reply spec/models/measure_spec.rb:119]: perfect - thanks!

File: spec/models/measure_spec.rb [hunk 4]
```diff
@@ -95,4 +95,49 @@
...
+        let(:user_id) { user_measure.user_id }
+
+        it "won't send when relationship_updated_at changes" do
+          expect { subject.touch(:relationship_updated_at) }
+            .not_to change { ActionMailer::Base.deliveries.count }.from(0)
+        end
+      end
+
+      context "when the current user doesn't have the user_measure" do
```
Comment (parndt) [inline spec/models/measure_spec.rb:128]:
```suggestion
      context "when the current user doesn't own the task" do
```

File: app/models/measure.rb (hunk 1 above)
Comment (parndt) [reply app/models/measure.rb:4]: we could call it `notifications` like we have on measures

**Comment by parndt** [comment]
@tmfrnz I added a field `notifications` to `Measuretype` defaulting to `false` which takes the place of the hardcoded task ID of 5

**Comment by parndt** [comment]
@tmfrnz I've merged in the latest code

File: app/controllers/measures_controller.rb [hunk 5]
```diff
@@ -35,9 +35,8 @@ def update
       return render json: '{"error":"Record outdated"}', status: :unprocessable_entity
     end
 
-    originally_draft = @measure.draft?
     if @measure.update!(permitted_attributes(@measure))
-      send_published_notification!(@measure) if originally_draft && !@measure.draft?
```
Comment (tmfrnz) [inline app/controllers/measures_controller.rb:40]: this should be retained!

File: app/controllers/measures_controller.rb [hunk 6]
```diff
@@ -63,16 +62,6 @@ def base_object
     end
   end
 
-  def send_published_notification!(measure)
```
Comment (tmfrnz) [inline app/controllers/measures_controller.rb:66]: likewise - this should be retained!

File: app/controllers/measures_controller.rb (hunk 5 above)
Comment (parndt) [reply app/controllers/measures_controller.rb:40]: great point @tmfrnz 👍

**Comment by parndt** [comment]
@tmfrnz took a look and fixed the code that got accidentally removed in a merge, and tidied up the tests related to that. Let me know what you think.

**Comment by tmfrnz** [comment]
@parndt thank you very much! Looks great at first glance - will test this in due course

**Comment by tmfrnz** [comment]
@parndt finally testing this and I am wondering if notifications are only set up to be sent on "relationship updates" or also updates to the measure record itself? my understanding of the code and tests suggests that it is only set up for the former but we do of course also need the latter

**Comment by tmfrnz** [comment]
> @parndt finally testing this and ~I am wondering if notifications are only set up to be sent on "relationship updates" or also updates to the measure record itself? my understanding of the code and tests suggests that it is only set up for the former but we do of course also need the latter~ while the "task updated" notifications are successfully sent when the task itself is updated (e.g. title, target_date), they are unfortunately not sent when changing a relationship. Interestingly the specs test only the latter and suggest it works.
> 
> My suspicion was that the "notify?" check of the measure model excludes changes to the `relationship_updated_at` and I have therefore tried adding the field to the list of notifiable attributes ([e203d89](https://github.com/dumparkltd/advocacy-tracker-server/commit/e203d89bc38f30e4ded29a7267a8b18eafddd69c)) and then to the "notify?" check itself ([1493960](https://github.com/dumparkltd/advocacy-tracker-server/commit/1493960d88986424c0f01dfb3c543238a383d1b1)) however without success

fixed by using update_attribute instead of update_column - the latter does not trigger callbacks https://github.com/dumparkltd/advocacy-tracker-server/pull/54/commits/bc31366b02011ce0aa49451458dd801ec4a1d241

**Pull request merged**